ui/app/src/elements/components/__tests__/render-block.test.ts

```diff
@@ -0,0 +1,109 @@
+import { property } from "lit/decorators.js"
+import { NHButton } from '@neighbourhoods/design-system-components'
+import { Renderer } from "@neighbourhoods/nh-launcher-applet"
+import { RenderBlock } from '../render-block'
+
+customElements.define('render-block', RenderBlock)
+
+// :TODO: This should really be wrapped up into a helper generator fn in some
+//        shared @neighbourhoods/applet-framework module.
+const renderer = (tagName, classDef) => (i: HTMLElement, r: CustomElementRegistry) => {
+  r.define(tagName, classDef)
+  i.innerHTML = `<${tagName} />`
+}
+
+describe('RenderBlock', () => {
+  const initialRender = async (theHTML) => {
+    return await fixture(theHTML)
+  }
+
+  describe('Given two components sharing a nested child CustomElement ', () =>
+  {
+    class Child1 extends ScopedRegistryHost(LitElement) {
+      static elementDefinitions = {
+        'nh-button': NHButton,
+      }
+
+      render() {
+        return html`<nh-button label="Child 1" />`
+      }
+    }
+
+    class Child2 extends ScopedRegistryHost(LitElement) {
+      static elementDefinitions = {
+        'nh-button': NHButton,
+      }
+
+      render() {
+        return html`<nh-button label="Child 2" />`
+      }
+    }
+
+    test(`They should be able to render same child elements`, async () => {
+      const harness = await initialRender(testHtml`<div>
+        <render-block .renderer=${renderer('child-1', Child1)}></render-block>
+        <render-block .renderer=${renderer('child-2', Child2)}></render-block>
+      </div>`)
+
+      const children = harness.querySelectorAll('render-block')
+
+      expect(children.length).to.equal(2)
+      expect(children[0].shadowRoot?.querySelector('child-1')).shadowDom.to.equal('<nh-button label="Child 1" />')
+      expect(children[1].shadowRoot?.querySelector('child-2')).shadowDom.to.equal('<nh-button label="Child 2" />')
+    })
+  })
+
+  describe('Given a RenderBlock rendering a child CustomElement with dynamic children ', () =>
+  {
+    class Child3 extends ScopedRegistryHost(LitElement) {
+      static elementDefinitions = {
+        'nh-button': NHButton,
+      }
+
+      render() {
+        return html`<nh-button label="Child 3" />`
+      }
+    }
+
+    class DynamicRenderingChild extends ScopedRegistryHost(LitElement) {
+      static elementDefinitions = {
+        'render-block': RenderBlock,
+        'child-3': Child3,
+      }
+
+      render() {
+        return html`<div>
+          <render-block .renderer=${renderer('child-3', Child3)}></render-block>
+          <child-3></child-3>
+        </div>`
+      }
+    }
+
+    test(`RenderBlock children should be able to be rendered recursively no matter their parent elements`, async () => {
+      const harness = await initialRender(testHtml`<div>
+        <render-block .renderer=${renderer('dynamic-child', DynamicRenderingChild)}></render-block>
+        <render-block .renderer=${renderer('child-3', Child3)}></render-block>
```

Comment:
@adaburrows I think we can probably mark this test as failing and leave it for later, because we'll want it to pass *eventually* when the Launcher becomes capable of mixing Resource types into different UI arrangements that may be dependent on each other (eg. image attachments to a post). But for now, maybe doesn't matter?

The error here occurs because `ScopedRegistryMixin` instantiates itself *statically*, meaning that all instances of the same CustomElement class share the same registry. This presents a problem when a RenderBlock may recursively render itself since all RenderBlocks use the same registry no matter their DOM nesting. In this case `Child3` is registered once on this line and once again (causing an error) when `dynamic-child` instantiates and re-registers `Child3` via its own `elementDefinitions`.

There may be a couple of ways of fixing this:

- One possibility is that RenderBlock could use its parent element's CustomElementRegistry? In this case elements would always be scoped to the registry of the parent element responsible for rendering the block. (In this test case, `Child3` would be registered with `global.customElements` here and with `DynamicRenderingChild.registry` above.)
- Another option would be to implement custom CustomElementRegistry handling within the RenderBlock class such that every *instance* gets its own registry rather than all instances sharing the same one.

To note also that none of this addresses potential name collisions in the dynamic tag names provided to RenderBlock's `renderer` but presumably this is managed by the Launcher performing some namespacing based on the app ID.

Reply:
Just to make a record of which way we went. We went ahead and overwrote the `createRenderRoot` to be a [little more like the mix-in](https://github.com/lit/lit/blob/08d48e3c31da07811b86f0aeae557e140b9c8c08/packages/labs/scoped-registry-mixin/src/scoped-registry-mixin.ts#L37), but with a instance associated `CustomElementRegistry`.